src/cpu/flexcpu/simple_dataflow_thread.hh

```diff
@@ -0,0 +1,620 @@
+ * met: redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer;
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Authors: Bradley Wang
+ */
+
+#ifndef __CPU_FLEXCPU_SIMPLE_DATAFLOW_THREAD_HH__
+#define __CPU_FLEXCPU_SIMPLE_DATAFLOW_THREAD_HH__
+
+#include <list>
+#include <memory>
+#include <set>
+#include <unordered_map>
+
+#include "cpu/exec_context.hh"
+#include "cpu/flexcpu/generic_reg.hh"
+#include "cpu/flexcpu/inflight_inst.hh"
+#include "cpu/flexcpu/simple_dataflow_cpu.hh"
+#include "cpu/inst_seq.hh"
+#include "cpu/reg_class.hh"
+#include "cpu/simple_thread.hh"
+#include "mem/request.hh"
+
+class SimpleDataflowCPU;
+
+/**
+ * Instances of this class represent a logical "Hyperthread" within the
+ * SimpleDataflowCPU. A thread contains an independent complete architectural
+ * state, and in this case, also the logical control flow.
+ */
+class SDCPUThread : public ThreadContext
+{
+  protected:
+    // BEGIN SDCPUThread Internal definitions
+
+    class MemIface : public InflightInst::MemIface
+    {
+      protected:
+        SDCPUThread& sdCPUThread;
+      public:
+        MemIface(SDCPUThread& sdCPUThread_):
+            sdCPUThread(sdCPUThread_)
+        { }
+
+        Fault initiateMemRead(std::shared_ptr<InflightInst> inst, Addr addr,
+                             unsigned int size, Request::Flags flags) override;
+
+        Fault writeMem(std::shared_ptr<InflightInst> inst, uint8_t *data,
+                       unsigned int size, Addr addr, Request::Flags flags,
+                       uint64_t *res) override;
+    };
+
+    struct SplitRequest {
+        RequestPtr main = nullptr;
+        RequestPtr low = nullptr;
+        RequestPtr high = nullptr;
+    };
```

Comment:
I don't think so, and here's why:

- The thread needs to know the physical address of all requests made, for tracking data dependencies through memory
- Split requests have two physical addresses, so the thread needs to hold enough information to complete both translations
- The main RequestPtr which will be owned by the PacketPtr passed to completeAcc() needs to be kept alive until that call can be made, otherwise it would have to be rebuilt, which I don't think is a better approach.